### FEATURES
The various features of the cleaned dataset are explained below:
1) Airline: The name of the airline company is stored in the airline column. It is a categorical feature having 6 different airlines.
2) Flight: Flight stores information regarding the plane's flight code. It is a categorical feature.
3) Source City: City from which the flight takes off. It is a categorical feature having 6 unique cities.
4) Departure Time: This is a derived categorical feature obtained created by grouping time periods into bins. It stores information about the departure time and have 6 unique time labels.
5) Stops: A categorical feature with 3 distinct values that stores the number of stops between the source and destination cities.
6) Arrival Time: This is a derived categorical feature created by grouping time intervals into bins. It has six distinct time labels and keeps information about the arrival time.
7) Destination City: City where the flight will land. It is a categorical feature having 6 unique cities.
8) Class: A categorical feature that contains information on seat class; it has two distinct values: Business and Economy.
9) Duration: A continuous feature that displays the overall amount of time it takes to travel between cities in hours.
10) Days Left: This is a derived characteristic that is calculated by subtracting the trip date by the booking date.
11) Price: Target variable stores information of the ticket price.

In [1]:
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt 
import numpy as np 
%matplotlib inline 

In [2]:
filepath = 'C:\\Python Folder\\My-Python\\Complete-Data-Science-With-Machine-Learning-And-NLP-2024-main\\Complete-Data-Science-With-Machine-Learning-And-NLP-2024-main\\Feature engineering\\flight_price.xlsx'

In [3]:
df = pd.read_excel(filepath)

In [4]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

#### Lets get the basic data info about the dataset

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [6]:
df.describe() # We have only one numerical feature here. i.e. Price 

Price
count  10683.000000
mean    9087.064121
std     4611.359167
min     1759.000000
25%     5277.000000
50%     8372.000000
75%    12373.000000
max    79512.000000

We will try to make this dataset ready for machine learning. I will go step by step to perform feature engineering which ensures that data is fully prepared

In [7]:
## Feature Engineering
# Lets get on with the date first 
df['Date'] = df['Date_of_Journey'].str.split('/').str[0]
df['Month'] = df['Date_of_Journey'].str.split('/').str[1]
df['Year'] = df['Date_of_Journey'].str.split('/').str[2]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
 11  Date             10683 non-null  object
 12  Month            10683 non-null  object
 13  Year             10683 non-null  object
dtypes: int64(1), object(13)
memory usage: 1.1+ MB


Converting dates from object type to int type, this is essential while making model learn. It is easier and faster for model to get data in int type 

In [9]:
df['Date'] = df['Date'].astype(int)
df['Month'] = df['Month'].astype(int)
df['Year'] = df['Year'].astype(int)

Since we already have dates, now dropping the original column

In [10]:
df.drop('Date_of_Journey', axis =1, inplace=True)
df.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

   Arrival_Time Duration Total_Stops Additional_Info  Price  Date  Month  Year  
0  01:10 22 Mar   2h 50m    non-stop         No info   3897    24      3  2019  
1         13:15   7h 25m     2 stops         No info   7662     1      5  2019  
2  04:25 10 Jun      19h     2 stops         No info  13882     9      6  2019  
3         23:30   5h 25m      1 stop         No info   6218    12      5  2019  
4         21:35   4h 45m      1 stop         No info  13302     1      3  2019

In [11]:
# Before moving to other time columns we need to clean some data from Arrival time 
df['Arrival_Time'] =df['Arrival_Time'].apply( lambda x:x.split(' ')[0])

I applied lambda function as some arrival dates have come with month in string format. Since we already have month column so no need of Month here

In [12]:
## Lets do same with other time columns. 
df['Arrival_hr'] = df['Arrival_Time'].str.split(':').str[0]
df['Arrival_min'] = df['Arrival_Time'].str.split(':').str[1]

In [13]:
df.head(2)

Airline    Source Destination                  Route Dep_Time  \
0     IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1  Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   

  Arrival_Time Duration Total_Stops Additional_Info  Price  Date  Month  Year  \
0        01:10   2h 50m    non-stop         No info   3897    24      3  2019   
1        13:15   7h 25m     2 stops         No info   7662     1      5  2019   

  Arrival_hr Arrival_min  
0         01          10  
1         13          15

In [14]:
df['Arrival_hr'] = df['Arrival_hr'].astype(int)
df['Arrival_min'] = df['Arrival_min'].astype(int)

In [15]:
df.drop('Arrival_Time', axis=1, inplace= True)
df.head(2)

Airline    Source Destination                  Route Dep_Time Duration  \
0     IndiGo  Banglore   New Delhi              BLR → DEL    22:20   2h 50m   
1  Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   7h 25m   

  Total_Stops Additional_Info  Price  Date  Month  Year  Arrival_hr  \
0    non-stop         No info   3897    24      3  2019           1   
1     2 stops         No info   7662     1      5  2019          13   

   Arrival_min  
0           10  
1           15

In [16]:
df['Departure_hr'] = df['Dep_Time'].str.split(':').str[0]
df['Departure_min'] = df['Dep_Time'].str.split(':').str[1]
df['Departure_hr'] = df['Departure_hr'].astype(int)
df['Departure_min'] = df['Departure_min'].astype(int)


In [17]:
df.drop('Dep_Time', axis=1 , inplace = True)
df.head(2)

Airline    Source Destination                  Route Duration  \
0     IndiGo  Banglore   New Delhi              BLR → DEL   2h 50m   
1  Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   

  Total_Stops Additional_Info  Price  Date  Month  Year  Arrival_hr  \
0    non-stop         No info   3897    24      3  2019           1   
1     2 stops         No info   7662     1      5  2019          13   

   Arrival_min  Departure_hr  Departure_min  
0           10            22             20  
1           15             5             50

Here we can have the duration is terms of minutes or hrs. I decided to make seperate columns for each of them

In [18]:
df['Dur_Hour']= df['Duration'].str.strip(' ').str[0]
df['Dur_Minutes'] = df['Duration'].str.split(' ').str[1].str.split('m').str[0]


Since there are some data points which only have hours , so making the minutes value to 0 

In [19]:
# Fill minutes with 0 instead on nan values
df['Dur_Minutes'].fillna('0', inplace=True)

In [20]:
df['Dur_Hour']= df['Dur_Hour'].astype(int)
df['Dur_Minutes'] = df['Dur_Minutes'].astype(int)

In [22]:
df.drop(df[['Duration']], axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Source           10683 non-null  object
 2   Destination      10683 non-null  object
 3   Route            10682 non-null  object
 4   Total_Stops      10682 non-null  object
 5   Additional_Info  10683 non-null  object
 6   Price            10683 non-null  int64 
 7   Date             10683 non-null  int32 
 8   Month            10683 non-null  int32 
 9   Year             10683 non-null  int32 
 10  Arrival_hr       10683 non-null  int32 
 11  Arrival_min      10683 non-null  int32 
 12  Departure_hr     10683 non-null  int32 
 13  Departure_min    10683 non-null  int32 
 14  Dur_Hour         10683 non-null  int32 
 15  Dur_Minutes      10683 non-null  int32 
dtypes: int32(9), int64(1), object(6)
memory usage: 959.9+ KB


In [23]:
df.head(2)

Airline    Source Destination                  Route Total_Stops  \
0     IndiGo  Banglore   New Delhi              BLR → DEL    non-stop   
1  Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR     2 stops   

  Additional_Info  Price  Date  Month  Year  Arrival_hr  Arrival_min  \
0         No info   3897    24      3  2019           1           10   
1         No info   7662     1      5  2019          13           15   

   Departure_hr  Departure_min  Dur_Hour  Dur_Minutes  
0            22             20         2           50  
1             5             50         7           25

#### Now we have columns [Airline], [Source], [Destination], [Total_stops], [Additional_info] to transform and get it ready for model learning

In [24]:
# Let deal with Total Stops and drop the route column
df.drop('Route', axis =1 , inplace= True)
df.Total_Stops.unique()  # Lets check the column content 

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

We have nan values here in categorical values. we can drop or replace it using Target guided ordinal encoding , in short filling it with mode value.

In [25]:
df[df['Total_Stops'].isnull()] # checking how many rows are affected

Airline Source Destination Total_Stops Additional_Info  Price  Date  \
9039  Air India  Delhi      Cochin         NaN         No info   7480     6   

      Month  Year  Arrival_hr  Arrival_min  Departure_hr  Departure_min  \
9039      5  2019           9           25             9             45   

      Dur_Hour  Dur_Minutes  
9039         2           40

In [26]:
df['Total_Stops'].mode()

0    1 stop
Name: Total_Stops, dtype: object

Hence we can use value = 1 for total stops in row 9039. 

In [27]:
# Since this cannot be treated with fillna method. we have to use map and assign manually, 
# as the prices may increase with number of stops so it is better if we rank this 
df['Total_Stops'] = df['Total_Stops'].map({'non-stop':0, '2 stops':2, '1 stop':1, '3 stops':3, np.nan:1, '4 stops':4})

In [28]:
# checking if any other nan left 
df[df['Total_Stops'].isnull()]

Empty DataFrame
Columns: [Airline, Source, Destination, Total_Stops, Additional_Info, Price, Date, Month, Year, Arrival_hr, Arrival_min, Departure_hr, Departure_min, Dur_Hour, Dur_Minutes]
Index: []

In [29]:
df.Additional_Info.value_counts()

Additional_Info
No info                         8345
In-flight meal not included     1982
No check-in baggage included     320
1 Long layover                    19
Change airports                    7
Business class                     4
No Info                            3
1 Short layover                    1
Red-eye flight                     1
2 Long layover                     1
Name: count, dtype: int64

In this feature found the duplicate 'No Info' and 'No info' which is literally same, so ensuring there are no other categorical variables with spaces, lowercases or similar case.

In [33]:
df[df['Additional_Info'] == 'No Info'] # 3 rows are affected 

Airline    Source Destination  Total_Stops Additional_Info  Price  Date  \
795   IndiGo  Banglore   New Delhi            1         No Info   8153     1   
2212  IndiGo  Banglore   New Delhi            1         No Info   7917     1   
9220  IndiGo  Banglore   New Delhi            1         No Info   8064     1   

      Month  Year  Arrival_hr  Arrival_min  Departure_hr  Departure_min  \
795       3  2019          21           50            16             55   
2212      3  2019          20           35            16             15   
9220      3  2019          10            0             5             15   

      Dur_Hour  Dur_Minutes  
795          4           55  
2212         4           20  
9220         4           45

In [41]:
df = df.replace('No Info', 'No info')

In [43]:
df.Additional_Info.unique() # Checking the unique values now 

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [53]:
df.Airline.unique() # 12 airlines are operating

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [45]:
df.Source.unique()  # There are 5 source originating airports

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [55]:
df.Destination.unique()  # There are 6 destinations

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [56]:
df.Total_Stops.unique() # Maximum 4 stops airlines are there. 

array([0, 2, 1, 3, 4], dtype=int64)

#### It is better to use Onehot Encoding here as each feature above can influence the prices of flight

In [57]:
# Using the One Hot Encoder for other features 
from sklearn.preprocessing import OneHotEncoder 
onehotenc = OneHotEncoder() # Instantiating onehotencoder

In [60]:
encoding = onehotenc.fit_transform(df[['Airline','Source','Destination', 'Additional_Info']]).toarray()

In [61]:
df_encoded = pd.DataFrame(data=encoding, columns=onehotenc.get_feature_names_out())

Airline    Source Destination  Total_Stops Additional_Info  Price  \
0       IndiGo  Banglore   New Delhi            0         No info   3897   
1    Air India   Kolkata    Banglore            2         No info   7662   
2  Jet Airways     Delhi      Cochin            2         No info  13882   
3       IndiGo   Kolkata    Banglore            1         No info   6218   
4       IndiGo  Banglore   New Delhi            1         No info  13302   

   Date  Month  Year  Arrival_hr  Arrival_min  Departure_hr  Departure_min  \
0    24      3  2019           1           10            22             20   
1     1      5  2019          13           15             5             50   
2     9      6  2019           4           25             9             25   
3    12      5  2019          23           30            18              5   
4     1      3  2019          21           35            16             50   

   Dur_Hour  Dur_Minutes  
0         2           50  
1         7           25  
2         1            0  
3         5           25  
4         4           45

In [62]:
df_encoded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 32 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   Airline_Air Asia                              10683 non-null  float64
 1   Airline_Air India                             10683 non-null  float64
 2   Airline_GoAir                                 10683 non-null  float64
 3   Airline_IndiGo                                10683 non-null  float64
 4   Airline_Jet Airways                           10683 non-null  float64
 5   Airline_Jet Airways Business                  10683 non-null  float64
 6   Airline_Multiple carriers                     10683 non-null  float64
 7   Airline_Multiple carriers Premium economy     10683 non-null  float64
 8   Airline_SpiceJet                              10683 non-null  float64
 9   Airline_Trujet                                10683 non-null 

Now merging this data with our df is the final step and the data is ready to be fed into machine learning model. Also dropping the columns encoded.

In [65]:
df_1 =pd.concat([df,df_encoded], axis=1)
df_1.head()

Airline    Source Destination  Total_Stops Additional_Info  Price  \
0       IndiGo  Banglore   New Delhi            0         No info   3897   
1    Air India   Kolkata    Banglore            2         No info   7662   
2  Jet Airways     Delhi      Cochin            2         No info  13882   
3       IndiGo   Kolkata    Banglore            1         No info   6218   
4       IndiGo  Banglore   New Delhi            1         No info  13302   

   Date  Month  Year  Arrival_hr  ...  Destination_New Delhi  \
0    24      3  2019           1  ...                    1.0   
1     1      5  2019          13  ...                    0.0   
2     9      6  2019           4  ...                    0.0   
3    12      5  2019          23  ...                    0.0   
4     1      3  2019          21  ...                    1.0   

   Additional_Info_1 Long layover  Additional_Info_1 Short layover  \
0                             0.0                              0.0   
1                             0.0                              0.0   
2                             0.0                              0.0   
3                             0.0                              0.0   
4                             0.0                              0.0   

   Additional_Info_2 Long layover  Additional_Info_Business class  \
0                             0.0                             0.0   
1                             0.0                             0.0   
2                             0.0                             0.0   
3                             0.0                             0.0   
4                             0.0                             0.0   

   Additional_Info_Change airports  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              0.0   

   Additional_Info_In-flight meal not included  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          0.0   
4                                          0.0   

   Additional_Info_No check-in baggage included  Additional_Info_No info  \
0                                           0.0                      1.0   
1                                           0.0                      1.0   
2                                           0.0                      1.0   
3                                           0.0                      1.0   
4                                           0.0                      1.0   

   Additional_Info_Red-eye flight  
0                             0.0  
1                             0.0  
2                             0.0  
3                             0.0  
4                             0.0  

[5 rows x 47 columns]

In [75]:
df_1.drop(columns= ['Airline','Source','Destination', 'Additional_Info'], inplace= True)
df_1.head(2)

Total_Stops  Price  Date  Month  Year  Arrival_hr  Arrival_min  \
0            0   3897    24      3  2019           1           10   
1            2   7662     1      5  2019          13           15   

   Departure_hr  Departure_min  Dur_Hour  ...  Destination_New Delhi  \
0            22             20         2  ...                    1.0   
1             5             50         7  ...                    0.0   

   Additional_Info_1 Long layover  Additional_Info_1 Short layover  \
0                             0.0                              0.0   
1                             0.0                              0.0   

   Additional_Info_2 Long layover  Additional_Info_Business class  \
0                             0.0                             0.0   
1                             0.0                             0.0   

   Additional_Info_Change airports  \
0                              0.0   
1                              0.0   

   Additional_Info_In-flight meal not included  \
0                                          0.0   
1                                          0.0   

   Additional_Info_No check-in baggage included  Additional_Info_No info  \
0                                           0.0                      1.0   
1                                           0.0                      1.0   

   Additional_Info_Red-eye flight  
0                             0.0  
1                             0.0  

[2 rows x 43 columns]

***DATA IS READY TO BE PROCESSED***